frame/contracts/src/wasm/prepare.rs

```diff
@@ -380,25 +359,80 @@ pub fn prepare_contract<C: ImportSatisfyCheck>(
 	} else {
 		// If none memory imported then just crate an empty placeholder.
 		// Any access to it will lead to out of bounds trap.
-		MemoryDefinition {
-			initial: 0,
-			maximum: 0,
-		}
-	};
+		Ok((0, 0))
+	}
+}
+
+/// Loads the given module given in `original_code`, performs some checks on it and
+/// does some preprocessing.
+///
+/// The checks are:
+///
+/// - provided code is a valid wasm module.
+/// - the module doesn't define an internal memory instance,
+/// - imported memory (if any) doesn't reserve more memory than permitted by the `schedule`,
+/// - all imported functions from the external environment matches defined by `env` module,
+///
+/// The preprocessing includes injecting code for gas metering and metering the height of stack.
+pub fn prepare_contract<C: ImportSatisfyCheck, T: Trait>(
+	original_code: &[u8],
+	schedule: &Schedule<T>,
+) -> Result<PrefabWasmModule, &'static str> {
+	let mut contract_module = ContractModule::new(original_code, schedule)?;
+	contract_module.scan_exports()?;
+	contract_module.ensure_no_internal_memory()?;
+	contract_module.ensure_table_size_limit(schedule.max_table_size)?;
+	contract_module.ensure_no_floating_types()?;
+
+	// We disallow importing `gas` function here since it is treated as implementation detail.
+	let disallowed_imports = [b"gas".as_ref()];
+	let memory_limits = get_memory_limits(
+		contract_module.scan_imports::<C>(&disallowed_imports)?,
+		schedule
+	)?;
 
 	contract_module = contract_module
 		.inject_gas_metering()?
 		.inject_stack_height_metering()?;
 
 	Ok(PrefabWasmModule {
 		schedule_version: schedule.version,
-		initial: memory_def.initial,
-		maximum: memory_def.maximum,
+		initial: memory_limits.0,
+		maximum: memory_limits.1,
 		_reserved: None,
 		code: contract_module.into_wasm_code()?,
 	})
 }
 
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking {
```

Comment:
Could you leave a comment why exactly we're doing this?

Reply:
Done.